client.go

```diff
@@ -0,0 +1,146 @@
+// Package checkout contains Go client for Checkout API.
+//
+// Checkout API documentation: https://docs.checkout.com/v2.0
+package checkout
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/pkg/errors"
+)
+
+// Caller makes HTTP call with given options and decode response into given struct.
+// Client implements this interface and pass itself to entity clients. You may create entity clients with own caller for
+// test purposes.
+type Caller interface {
+	Call(ctx context.Context, method, path string, headers map[string]string, reqObj interface{}, respObj interface{}) error
+}
+
+// HTTPClient is interface fot HTTP client. Built-in net/http.Client implements this interface as well.
+type HTTPClient interface {
+	Do(r *http.Request) (*http.Response, error)
+}
+
+// Option is a callback for redefine client parameters.
+type Option func(*Client)
+
+// Client contains API parameters and provides set of API entity clients.
+type Client struct {
+	httpClient HTTPClient
+	secretKey  string
+}
+
+const (
+	apiURL = "https://api.checkout.com/"
+
+	headerAuthorization = "Authorization"
+	headerIdempotency   = "Cko-Idempotency-Key"
+)
+
+// New creates new client with given options.
+func New(options ...Option) *Client {
+	c := &Client{
+		httpClient: http.DefaultClient,
+	}
+
+	for _, option := range options {
+		option(c)
+	}
+
+	return c
+}
+
+// OptHTTPClient returns option with given HTTP client.
+func OptHTTPClient(httpClient HTTPClient) Option {
+	return func(c *Client) {
+		c.httpClient = httpClient
+	}
+}
+
+// OptSecretKey returns option with given secret key.
+func OptSecretKey(secretKey string) Option {
+	return func(c *Client) {
+		c.secretKey = secretKey
+	}
+}
+
+// Call does HTTP request with given params using set HTTP client. Response will be decoded into respObj.
+// CallError may be returned if something went wrong. If API return error as response, then Call returns error of type checkout.CallError.
+func (c *Client) Call(ctx context.Context, method, path string, idempotencyKey string, reqObj interface{}, respObj interface{}) (callErr error) {
+	var reqBody io.Reader
+
+	if reqObj != nil {
+		reqBodyBytes, err := json.Marshal(reqObj)
+		if err != nil {
+			return errors.Wrap(err, "failed to marshal request body")
+		}
+		reqBody = bytes.NewBuffer(reqBodyBytes)
+	}
+
+	req, err := http.NewRequest(method, apiURL+path, reqBody)
+	if err != nil {
+		return errors.Wrap(err, "failed to create HTTP request")
+	}
+
+	req = req.WithContext(ctx)
+
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set(headerAuthorization, c.secretKey)
+
+	if idempotencyKey != "" {
```

Comment:
Checkout strongly recommends V4 UUID, so may be just generate UUID from request data with `func NewSHA1(space UUID, data []byte) UUID` from UUID package?

Reply:
I want that this key is managed by client consumer. What you think if client will generate UUID from string value which is defined by consumer?

Reply:
I am ok with current approach, instead of cases when idempotencyKey is empty, this gives a chance for double transactions etc.

Reply:
It should be managed by consumer, not by our client. Idempotency key is optional in API doc.